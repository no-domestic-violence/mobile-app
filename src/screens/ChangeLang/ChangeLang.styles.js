import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  userSettingsContainer: {
    alignItems: 'flex-start',
  },
  arrow: {
    alignSelf: 'flex-start',
    marginTop: 50,
    marginLeft: 10,
  },
  button: {
    margin: 20,
    alignItems: 'center',
  },
  buttonText: {
    backgroundColor: '#FFFFFF',
    paddingVertical: 20,
    paddingHorizontal: 90,
    borderRadius: 50,
    alignItems: 'center',
    justifyContent: 'center',
  },
  view: {
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
  },
  header: {
    fontSize: 20,
    fontWeight: '600',
    alignSelf: 'flex-start',
    marginLeft: 30,
    marginBottom: 40,
    marginTop: 40,
  },
});
