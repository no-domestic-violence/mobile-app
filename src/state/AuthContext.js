import AsyncStorage from '@react-native-async-storage/async-storage';
import createAppContext from './CreateAppContext';
import appApiClient from '../api/appApiClient';

const authReducer = (state, action) => {
  switch (action.type) {
    case 'SIGNUP_ERROR':
    case 'LOGIN_ERROR':
      return {
        ...state,
        errorMessage: action.payload,
      };
    case 'SIGNUP_SUCCESS':
    case 'LOGIN_SUCCESS':
      return {
        ...state,
        token: action.payload.token,
        username: action.payload.user.username,
        errorMessage: '',
      };
    case 'CHANGE_PASSWORD_SUCCESS':
      return {
        ...state,
        successMessage: action.payload,
      };
    case 'CHANGE_PASSWORD_ERROR':
      return {
        ...state,
        errorMessage: action.payload,
      };
    case 'AUTH_SUCCESS':
      return {
        ...state,
        token: action.payload.token,
        username: action.payload.username,
      };
    case 'LOGOUT':
      return {
        token: '',
        username: '',
      };
    case 'DELETE_ACCOUNT':
      return {
        token: '',
        username: '',
        successMessage: action.payload.data,
      };
    case 'REMOVE_ERRORS':
      return {
        ...state,
        errorMessage: '',
      };
    case 'REMOVE_MESSAGES':
      return {
        ...state,
        successMessage: '',
      };
    default:
      return state;
  }
};

const signup = (dispatch) => async ({ email, password, username }) => {
  try {
    const response = await appApiClient.post('/signup', {
      email,
      password,
      username,
    });

    await AsyncStorage.setItem('token', response.data.token);
    await AsyncStorage.setItem('username', response.data.user.username);

    dispatch({ type: 'SIGNUP_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'SIGNUP_ERROR',
      payload: "Something went wrong'( Try again",
    });
  }
};

const login = (dispatch) => async ({ email, password }) => {
  try {
    const response = await appApiClient.post('/login', { email, password });
    await AsyncStorage.setItem('token', response.data.token);
    await AsyncStorage.setItem('username', response.data.user.username);
    dispatch({ type: 'LOGIN_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'LOGIN_ERROR',
      payload: 'Are you sure about password and email?',
    });
  }
};

const authentication = (dispatch) => async () => {
  const token = await AsyncStorage.getItem('token');
  const username = await AsyncStorage.getItem('username');
  const user = { token, username };
  if (token) {
    dispatch({ type: 'AUTH_SUCCESS', payload: user });
  }
};
const removeErrors = (dispatch) => () => {
  dispatch({ type: 'REMOVE_ERRORS' });
};

const removeMessages = (dispatch) => () => {
  dispatch({ type: 'REMOVE_MESSAGES' });
};

const changePassword = (dispatch) => async ({
  email,
  password,
  oldPassword,
}) => {
  try {
    const response = await appApiClient.post('/changePassword', {
      email,
      oldPassword,
      password,
    });
    removeErrors();
    dispatch({ type: 'CHANGE_PASSWORD_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'CHANGE_PASSWORD_ERROR',
      payload: error.response.data.message,
    });
  }
};

const signout = (dispatch) => async () => {
  await AsyncStorage.removeItem('token');
  dispatch({ type: 'LOGOUT' });
};

const deleteAccount = (dispatch) => async ({ username }) => {
  try {
    // TODO: fix sending params so it is more readable
    const response = await appApiClient.delete('/deleteUser', {
      params: { username },
    });
    await AsyncStorage.removeItem('token');
    await AsyncStorage.removeItem('username');
    dispatch({ type: 'DELETE_ACCOUNT', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'CHANGE_PASSWORD_ERROR',
      payload: error.response.data.message,
    });
  }
};

export const { Provider, Context } = createAppContext(
  authReducer,
  {
    signup,
    login,
    signout,
    removeErrors,
    removeMessages,
    authentication,
    changePassword,
    deleteAccount,
  },
  { isLoggedIn: false, errorMessage: '' }
);
