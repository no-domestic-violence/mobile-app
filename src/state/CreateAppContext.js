import React, { useReducer } from 'react';

// boilerplate for creating Context and Provider
// does not to be changed
export default (reducer, actions, defaultValue) => {
  const Context = React.createContext();

  const Provider = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, defaultValue);

    //actions is object for ex {updateCurrentLocation: (dispatch)=> (location) => { dispatch({ type: 'UPDATE_LOCATION', payload: location });}
    const boundActions = {};
    //we need to loop through each action in object and for every key(which is function) we call it with dispatch argument and it gives back function. this function we pass down to all children like ...boundActions
    for (const key in actions) {
      boundActions[key] = actions[key](dispatch);
    }
    return (
      <Context.Provider value={{ state, ...boundActions }}>
        {children}
      </Context.Provider>
    );
  };
  return { Context, Provider };
};
