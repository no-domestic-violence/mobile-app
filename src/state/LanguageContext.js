import React, { createContext, useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import i18n from 'services/i18n';
import PropTypes from 'prop-types';
import { DEFAULT_LANGUAGE } from '../services/i18n/language-detector';

const APP_LANGUAGE = 'lang';

const availableLanguages = [
  { lang: 'en', name: 'English' },
  { lang: 'de', name: 'German' },
];

export const LanguageContext = createContext({
  availableLanguages,
  appLanguage: DEFAULT_LANGUAGE,
  setAppLanguage: () => {},
  initializeAppLanguage: () => {},
});

export const LanguageProvider = ({ children }) => {
  const [appLanguage, setAppLanguage] = useState(DEFAULT_LANGUAGE);

  const setLanguage = async (language) => {
    i18n.changeLanguage(language);
    setAppLanguage(language);
    AsyncStorage.setItem(APP_LANGUAGE, language);
  };

  const initializeAppLanguage = async () => {
    const currentLanguage = await AsyncStorage.getItem(APP_LANGUAGE);
    if (!currentLanguage) {
      setLanguage(DEFAULT_LANGUAGE);
    } else {
      setLanguage(currentLanguage);
    }
  };

  return (
    <LanguageContext.Provider
      value={{
        availableLanguages,
        setAppLanguage: setLanguage,
        appLanguage,
        initializeAppLanguage,
      }}>
      {children}
    </LanguageContext.Provider>
  );
};
LanguageProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
