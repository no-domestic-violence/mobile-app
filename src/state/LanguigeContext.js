//TODO: it did not work, to figure out if we need or we can just you provider for everything


// import React, { createContext } from 'react';
// import { I18nextProvider } from 'react-i18next';
// import i18next from 'i18next';

// const LanguigeContext = createContext();

// function LanguigeProvider(props) {

//   return (
//     <LanguigeContext.Provider>
//       <I18nextProvider i18n={i18next}>{props.children}</I18nextProvider>
//    </LanguigeContext.Provider>
//   );
// }

// export {LanguigeProvider, LanguigeContext};

