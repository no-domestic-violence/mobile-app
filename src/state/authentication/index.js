import createAppContext from 'helpers/createAppContext';
import {
  signup,
  login,
  signout,
  removeErrors,
  removeMessages,
  authentication,
  changePassword,
  deleteAccount,
  setAlreadyLaunchedValue,
} from './authentication.actions';
import authReducer from './authentication.reducer';

export const { Provider, Context } = createAppContext(
  authReducer,
  {
    signup,
    login,
    signout,
    removeErrors,
    removeMessages,
    authentication,
    changePassword,
    deleteAccount,
    setAlreadyLaunchedValue,
  },
  { isLoggedIn: false, errorMessage: '', isFirstLaunch: null }
);
