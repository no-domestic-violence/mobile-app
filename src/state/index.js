import React, {createContext, useState, useMemo} from 'react';

const AppContext = createContext();

function ContextProvider(props) {
  //just an example of context usage
  const [userToken, setUserToken] = useState('');
  const authContext = useMemo(() => {
    return {
      logIn: () => {
        setUserToken('tokenishere');
      },
      signUp: () => {
        setUserToken('tokenishere');
      },
      signOut: () => {
        setUserToken('');
      },
    };
  });

  return (
    <AppContext.Provider value={{authContext, userToken}}>
      {props.children}
    </AppContext.Provider>
  );
}

export {ContextProvider, AppContext};
