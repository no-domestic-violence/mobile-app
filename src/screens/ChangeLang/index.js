import React, { useContext } from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { Context as AuthContext } from '../../state/AuthContext';
import { StyledView } from '../../styles/shared/StyledView';
import UserInfo from '_components/user-settings/';
import { faAngleLeft } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import { LanguageContext } from '../../state/LanguageContext';
import { Button } from 'react-native-elements';

export default function ChangeLanguage({ navigation }) {
  const { state } = useContext(AuthContext);
  const { availableLanguages, setAppLanguage } = useContext(LanguageContext);
  const username = state.username;

  const handleLanguageChange = async (language) => {
    await setAppLanguage(language);
    navigation.goBack();
  };

  return (
    <StyledView style={styles.userSettingsContainer}>
      <UserInfo username={username} />
      <View style={styles.view}>
        <FontAwesomeIcon
          onPress={() => navigation.goBack()}
          icon={faAngleLeft}
          size={40}
          color={'#000'}
          style={styles.arrow}
        />
        <Text style={styles.header}>Change Language</Text>
        <View style={styles.buttonsView}>
          {availableLanguages.map((item, i) => (
            <Button
              title={item.name}
              titleStyle={{ color: '#000000' }}
              type="solid"
              raised
              containerStyle={styles.button}
              buttonStyle={styles.buttonText}
              key={i}
              onPress={() => handleLanguageChange(item.lang)}
            />
          ))}
        </View>
      </View>
    </StyledView>
  );
}

const styles = StyleSheet.create({
  userSettingsContainer: {
    alignItems: 'flex-start',
  },
  arrow: {
    alignSelf: 'flex-start',
    marginTop: 50,
    marginLeft: 10,
  },
  button: {
    margin: 20,
    alignItems: 'center',
  },
  buttonText: {
    backgroundColor: '#FFFFFF',
    paddingVertical: 20,
    paddingHorizontal: 90,
    borderRadius: 50,
    alignItems: 'center',
    justifyContent: 'center',
  },
  view: {
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
  },
  header: {
    fontSize: 20,
    fontWeight: '600',
    alignSelf: 'flex-start',
    marginLeft: 30,
    marginBottom: 40,
    marginTop: 40,
  },
});
