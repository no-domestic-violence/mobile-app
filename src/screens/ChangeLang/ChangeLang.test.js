import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { AuthContext, LanguageContext } from 'state';
import ChangeLang from './index';

const state = {
    username : 'testuser'
}

const mockNavigation = {
    goBack: jest.fn()
}

const setAppLanguage = jest.fn();
const availableLanguages = [
    { name: 'english', lang: 'en'},
    {name: 'german', lang: 'de'}
]

describe('<ChangeLang />', () => {
    let screen;
    beforeEach(() => {
        screen = render(
            <LanguageContext.Provider value={{ setAppLanguage, availableLanguages }}>
               <AuthContext.Provider value={{ state }}>
                   <ChangeLang navigation={mockNavigation}/>
                   </AuthContext.Provider>
            </LanguageContext.Provider>);

    })
  it('matches snapshot', () => {
      screen.toJSON();
      expect(screen).toMatchSnapshot();
  });
    it('renders username from the state', () => {
        expect(screen.getByText(state.username)).toBeTruthy();
  })
    it('renders all the available language options', () => {
        expect(screen.getByText(/english/i)).toBeTruthy();
        expect(screen.getByText(/german/i)).toBeTruthy();
    });

    it('handles selected language change', async () => {
        await fireEvent.press(screen.getByText(/german/i));
        expect(setAppLanguage).toHaveBeenCalledWith('de');
        expect(mockNavigation.goBack).toHaveBeenCalledTimes(1);
    })
    
});
