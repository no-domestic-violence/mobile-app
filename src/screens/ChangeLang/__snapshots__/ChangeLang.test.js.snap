// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangeLang /> matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#cadeee",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "alignItems": "flex-start",
      },
    ]
  }
>
  <View>
    <svg
      data-file-name="SvgUserSettings"
      style={
        Object {
          "position": "absolute",
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginLeft": 10,
          "marginTop": 10,
        }
      }
    >
      <
        color="grey"
        icon={
          Object {
            "icon": Array [
              496,
              512,
              Array [],
              "f2bd",
              "M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z",
            ],
            "iconName": "user-circle",
            "prefix": "fas",
          }
        }
        size={80}
      />
      <Text
        style={
          Object {
            "fontSize": 28,
            "marginLeft": 15,
            "textTransform": "capitalize",
          }
        }
      >
        testuser
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <
      color="#000"
      icon={
        Object {
          "icon": Array [
            256,
            512,
            Array [],
            "f104",
            "M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z",
          ],
          "iconName": "angle-left",
          "prefix": "fas",
        }
      }
      onPress={[Function]}
      size={40}
      style={
        Object {
          "alignSelf": "flex-start",
          "marginLeft": 10,
          "marginTop": 50,
        }
      }
    />
    <Text
      style={
        Object {
          "alignSelf": "flex-start",
          "fontSize": 20,
          "fontWeight": "600",
          "marginBottom": 40,
          "marginLeft": 30,
          "marginTop": 40,
        }
      }
    >
      Change Language
    </Text>
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 50,
            "margin": 20,
            "overflow": "visible",
            "shadowColor": "rgba(0,0,0, .4)",
            "shadowOffset": Object {
              "height": 1,
              "width": 1,
            },
            "shadowOpacity": 1,
            "shadowRadius": 1,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderColor": "#2089dc",
                "borderRadius": 50,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
                "paddingHorizontal": 90,
                "paddingVertical": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              english
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 50,
            "margin": 20,
            "overflow": "visible",
            "shadowColor": "rgba(0,0,0, .4)",
            "shadowOffset": Object {
              "height": 1,
              "width": 1,
            },
            "shadowOpacity": 1,
            "shadowRadius": 1,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderColor": "#2089dc",
                "borderRadius": 50,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
                "paddingHorizontal": 90,
                "paddingVertical": 20,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              german
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
